packages/contracts/src/contracts/current/forwarder/MixinForwarderQuote.sol

```diff
@@ -0,0 +1,139 @@
+        public
+        view
+        returns (Exchange.FillResults memory fillResults)
+    {
+        // Compute the order hash
+        bytes32 orderHash = getOrderHash(order);
+        uint256 remainingTakerAssetAmount = safeSub(order.takerAssetAmount, EXCHANGE.getUnavailableTakerTokenAmount(orderHash));
+        // Validate order expiration
+        if (block.timestamp >= order.expirationTimeSeconds) {
+            return fillResults;
+        }
+        // Validate order availability
+        fillResults.takerAssetFilledAmount = min256(takerAssetFillAmount, remainingTakerAssetAmount);
+        if (fillResults.takerAssetFilledAmount == 0) {
+            return fillResults;
+        }
+        // Validate fill order rounding
+        if (isRoundingError(fillResults.takerAssetFilledAmount, order.takerAssetAmount, order.makerAssetAmount)) {
+            fillResults.takerAssetFilledAmount = 0;
+            return fillResults;
+        }
+        fillResults.makerAssetFilledAmount = getPartialAmount(fillResults.takerAssetFilledAmount, order.takerAssetAmount, order.makerAssetAmount);
+        if (order.feeRecipientAddress != address(0)) {
+            if (order.makerFee > 0) {
+                fillResults.makerFeePaid = getPartialAmount(fillResults.takerAssetFilledAmount, order.takerAssetAmount, order.makerFee);
+            }
+            if (order.takerFee > 0) {
+                fillResults.takerFeePaid = getPartialAmount(fillResults.takerAssetFilledAmount, order.takerAssetAmount, order.takerFee);
+            }
+        }
+        return fillResults;
+    }
+
+    /// @dev Calculates a total for selling takerTokenFillAmount across all orders. Including the fees
+    ///      required to be paid. 
+    /// @param orders An array of Order struct containing order specifications.
+    /// @param takerAssetFillAmount A number representing the amount of this order to fill.
+    /// @return Amounts filled and fees paid by maker and taker.
+    function expectedMaketSellFillResults(Order[] memory orders, uint256 takerAssetFillAmount)
+        public
+        view
+        returns (Exchange.FillResults memory fillResult)
+    {
+        for (uint256 i = 0; i < orders.length; i++) {
+            require(areBytesEqual(orders[i].makerAssetData, orders[0].makerAssetData), "all orders must be the same token pair");
+            uint256 remainingTakerAssetFillAmount = safeSub(takerAssetFillAmount, fillResult.takerAssetFilledAmount);
+
+            Exchange.FillResults memory quoteFillResult = fillOrderQuote(orders[i], remainingTakerAssetFillAmount);
+
+            addFillResults(fillResult, quoteFillResult);
+            if (fillResult.takerAssetFilledAmount == takerAssetFillAmount) {
+                break;
+            }
+        }
+        return fillResult;
+    }
+
+    /// @dev Calculates a quote total for buyTokens. This is useful for off-chain queries to 
+    ///      ensure all calculations are performed atomically for consistent results
```

Comment:
Hm, doing this off-chain however is not guarenteed to give the same result as when the tx is actually submitted b/c of race conditions. When and how exactly do you expect this to be used?

Reply:
This is useful off-chain and the atomicity guarantee is for the marketSell and fee abstraction. It can be calculated off-chain though performing it in 2 separate calls could be spread over different block numbers. This is here to aid that situation. It's use would be an accurate display to the end user.

Reply:
Not sure I understand your explanation. Could you elaborate?

Reply:
The underlying action here is multiple steps, buy tokens to pay fees and then buy the requested tokens. 

We could query those separately with their respective `calculateExpectedFillResult` in two calls to the blockchain. During this time, between the two calls, the blockchain progresses to the next block where one has been partially filled. Let's say it's the REP orders where a partial fill occurred. We would be showing the user a less accurate quote in the UI.

The rate will be accurate to a single block rather than potentially over multiple.